bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -399,6 +399,42 @@ private static enum State {
         ACTIVE;
     }
 
+    /**
+     * Class for throttling HTTP GET and PUT requests to prevent overloading the Hue bridge.
+     * <p>
+     * The Hue Bridge can get confused if they receive too many HTTP requests in a short period of time (e.g. on start
+     * up), or if too many HTTP sessions are opened at the same time, which cause it to respond with an HTML error page.
+     * So this class a) waits to acquire permitCount (or no more than MAX_CONCURRENT_SESSIONS) stream permits, and b)
+     * throttles the requests to a maximum of one per REQUEST_INTERVAL_MILLISECS.
+     */
+    private class Throttler implements AutoCloseable {
+        private final int permitCount;
+
+        /**
+         * @param permitCount indicates how many stream permits to be acquired.
+         * @throws InterruptedException
+         */
+        Throttler(int permitCount) throws InterruptedException {
+            this.permitCount = permitCount;
+            streamMutex.acquire(permitCount);
+            long delay;
+            synchronized (Clip2Bridge.this) {
+                Instant now = Instant.now();
+                delay = lastRequestTime.map(t -> Duration.between(now, t).toMillis() + REQUEST_INTERVAL_MILLISECS)
+                        .orElse(0L);
+                lastRequestTime = Optional.of(delay > 0 ? now.plusMillis(delay) : now);
+            }
+            if (delay > 0) {
+                Thread.sleep(delay);
```

Comment:
Didn't you say this needs to be inside the synchronized block?

Reply:
I did. If the sleep would be inside the sychronized block then lastRequestTime must be set to Instant.now() after the sleep has expired. However the new formulation of lastRequestTime at line #425 means that subsequent calls rather than waiting on the sychronized block, will instead wait on a longer sleep. Honestly I don't know if blocking on synschronized or blocking on sleep is better code; I think probably the latter because it can be interrupted.

Reply:
Makes sense. I guess the difference is negligible, because interrupting the sleep also means giving up the lock and thus unlocking other waiting threads, but I also prefer making threads sleep over making them wait for locks.